### PR TITLE
Split encodeToMemory into packedEncode and abiEncode

### PR DESCRIPTION
**Pull request by axic**

### CODE REVIEW

File: libsolidity/codegen/ExpressionCompiler.cpp [hunk 1]
```diff
@@ -675,7 +675,7 @@ bool ExpressionCompiler::visit(FunctionCall const& _functionCall)
 				argumentTypes.push_back(arg->annotation().type);
 			}
 			utils().fetchFreeMemoryPointer();
-			utils().encodeToMemory(argumentTypes, TypePointers(), function.padArguments(), true);
+			utils().packedEncode(argumentTypes, TypePointer());
```
Comment (axic) [inline libsolidity/codegen/ExpressionCompiler.cpp]: Typo: TypePointers vs TypePointer

File: libsolidity/codegen/ExpressionCompiler.cpp [hunk 2]
```diff
@@ -675,7 +675,7 @@ bool ExpressionCompiler::visit(FunctionCall const& _functionCall)
 				argumentTypes.push_back(arg->annotation().type);
 			}
 			utils().fetchFreeMemoryPointer();
-			utils().encodeToMemory(argumentTypes, TypePointers(), function.padArguments(), true);
```
Comment (chriseth) [inline libsolidity/codegen/ExpressionCompiler.cpp:678]: Please add `solAssert(!function.padArguments(), "");`

**Comment by chriseth** [comment]
Good apart from the two minor changes.

File: libsolidity/codegen/ExpressionCompiler.cpp (hunk 2 above)
Comment (axic) [reply libsolidity/codegen/ExpressionCompiler.cpp:678]: It is moot I think. The only reason for `padArguments` to exists is to be used in `appendExternalFunctionCall`, though `SHA3` cannot be used in that context.

Comment (axic) [reply libsolidity/codegen/ExpressionCompiler.cpp:678]: Also it is used for a warning of inferring types. I'd actually rename it `usesPackedEncoding` or `usesTightPacking`.

Though that is the next step of refactoring making sense of this :)

**Comment by axic** [comment]
@chriseth updated

File: libsolidity/codegen/ExpressionCompiler.cpp (hunk 2 above)
Comment (chriseth) [reply libsolidity/codegen/ExpressionCompiler.cpp:678]: Still, at some point we might change the value and assume that this also changes the behaviour. Better more than fewer assertions.

**Review by chriseth — APPROVED** [review]

**Pull request merged**